dispatcher.go

```diff
@@ -323,176 +320,90 @@ func (d *Dispatcher) Register(rs []transport.Procedure) {
-	// might result in an outbound call before the outbound is ready.
-
-	var (
-		mu         sync.Mutex
-		allStarted []transport.Lifecycle
-	)
-
-	d.log.Info("Starting up.")
-	start := func(s transport.Lifecycle) func() error {
-		return func() error {
-			if s == nil {
-				return nil
-			}
-
-			if err := s.Start(); err != nil {
-				return err
-			}
-
-			mu.Lock()
-			allStarted = append(allStarted, s)
-			mu.Unlock()
-			return nil
-		}
+	starter := &PhasedStarter{
+		dispatcher: d,
+		log:        d.log,
 	}
-
-	abort := func(errs []error) error {
-		// Failed to start so stop everything that was started.
-		wait := errorsync.ErrorWaiter{}
-		for _, s := range allStarted {
-			wait.Submit(s.Stop)
+	return d.once.Start(func() error {
+		d.log.Info("starting dispatcher")
+		starter.setRouters()
+		if err := starter.StartTransports(); err != nil {
+			return err
 		}
-		if newErrors := wait.Wait(); len(newErrors) > 0 {
-			errs = append(errs, newErrors...)
+		if err := starter.StartOutbounds(); err != nil {
+			return err
 		}
+		if err := starter.StartInbounds(); err != nil {
+			return err
+		}
+		d.log.Info("dispatcher startup complete")
+		return nil
+	})
+}
 
-		return multierr.Combine(errs...)
-	}
-
-	// Set router for all inbounds
-	for _, i := range d.inbounds {
-		i.SetRouter(d.table)
-	}
-	d.log.Debug("Set router for inbounds.")
-
-	// Start Transports
-	wait := errorsync.ErrorWaiter{}
-	d.log.Debug("Starting transports.")
-	for _, t := range d.transports {
-		wait.Submit(start(t))
-	}
-	if errs := wait.Wait(); len(errs) != 0 {
-		return abort(errs)
-	}
-	d.log.Debug("Started transports.")
-
-	// Start Outbounds
-	wait = errorsync.ErrorWaiter{}
-	d.log.Debug("Starting outbounds.")
-	for _, o := range d.outbounds {
-		wait.Submit(start(o.Unary))
-		wait.Submit(start(o.Oneway))
-		wait.Submit(start(o.Stream))
-	}
-	if errs := wait.Wait(); len(errs) != 0 {
-		return abort(errs)
-	}
-	d.log.Debug("Started outbounds.")
-
-	// Start Inbounds
-	wait = errorsync.ErrorWaiter{}
-	d.log.Debug("Starting inbounds.")
-	for _, i := range d.inbounds {
-		wait.Submit(start(i))
-	}
-	if errs := wait.Wait(); len(errs) != 0 {
-		return abort(errs)
+// PhasedStart is a more granular alternative to Start, and is intended only
+// for advanced users. Rather than starting all transports, inbounds, and
+// outbounds at once, it lets the user start them separately.
+//
+// Start and PhasedStart are mutually exclusive. If Start is called first,
+// PhasedStart is a no-op and returns the same error (if any) that Start
+// returned. If PhasedStart is called first, Start is a no-op and always
+// returns a nil error; the caller is responsible for using the PhasedStarter
+// to complete startup.
+func (d *Dispatcher) PhasedStart() (*PhasedStarter, error) {
+	starter := &PhasedStarter{
+		dispatcher: d,
+		log:        d.log,
 	}
-	d.log.Debug("Started inbounds.")
-
-	d.log.Info("Started up.")
-	return nil
+	err := d.once.Start(func() error {
```

Comment:
nit: instead of always returning the starter, should we do a check here so that we can return a `nil` `PhasedStarter` on errors?

Reply:
Sure, sounds good.

Reply:
Added this bit of logic in d19718a5.